modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cluster/ClientClusterGroupProjection.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.platform.client.cluster;
+
+import org.apache.ignite.binary.BinaryRawReader;
+import org.apache.ignite.cluster.ClusterGroup;
+
+import java.util.function.Function;
+
+/**
+ * Client cluster group projection representation.
+ * Decodes a remote projection request from a client node.
+ */
+public class ClientClusterGroupProjection {
+    /** */
+    private static final short ATTRIBUTE = 1;
+
+    /** */
+    private static final short SERVER_NODES = 2;
+
+    /** Cluster group factory method. */
+    private final Function<ClusterGroup, ClusterGroup> clusterGrpFactory;
+
+    /**
+     * Constructor.
+     *
+     * @param clusterGrpFactory Cluster Group builder.
+     */
+    private ClientClusterGroupProjection(Function<ClusterGroup, ClusterGroup> clusterGrpFactory){
+        this.clusterGrpFactory = clusterGrpFactory;
+    }
+
+    /**
+     * Reads projection from a stream.
+     * @param reader Reader.
+     * @return Projection.
+     */
+    public static ClientClusterGroupProjection read(BinaryRawReader reader) {
+        if (!reader.readBoolean())
+            return new ClientClusterGroupProjection(null);
+
+        Function<ClusterGroup, ClusterGroup> factory = clusterGrp ->
```

Comment:
This function closes over `reader`, postponing actual stream reading until `apply` us called. This assumes that no one reads from that reader between `read` and `apply` calls, which works for now, but is fragile and will be hard to debug in case it fails in future.

Throughout our code base, `read` methods are supposed to read their share of data right away. Please refactor accordingly.

I think we can use functional composition here, something like:
```
            Function<ClusterGroup, ClusterGroup> factory = c -> c;

            int cnt = reader.readInt();
            for (int i = 0; i < cnt; i++) {
                short code = reader.readShort();
                final Function<ClusterGroup, ClusterGroup> prev = factory;
                switch (code) {
                    case ATTRIBUTE: {
                        String attName = reader.readString();
                        String attrVal = reader.readString();
                        factory = c -> prev.apply(c).forAttribute(attrName, attrVal);
                        break;
                    }
              ...
              return new ClientClusterGroupProjection(factory);
```

So the code is very similar, but reading is done right away.

Reply:
Refactored in .NET way with internal classes.